### DISCUSSION TITLE
자동차 경주 게임 피드백 및 생각해 볼만한 것

### DISCUSSION THREAD

**Discussion by OzRagwort** [woowa-Effective-Java-Study/effective-java-study#1] [Show and tell]
이전 미션하면서 피드백 받은 것과 고민했던 것들이에요

여러분들도 같이 고민해보고 이야기해봐요!

## 피드백 정리 + 생각해볼 만한 것들

- google java style 메소드 명은 동사로 시작해야 한다.
- toString()를 어떻게 잘 사용해야할지 고민하자!
    - toString()으로 View에 출력값을 전달하는게 맞는가?
        - 출력 형식이 바뀌어야 한다면 View와 Cars중 어느곳이 수정되어야 하는가?
        - VO가 View의 세부사항을 알고 있는것이 바람직한가?
- 1단계 페어 끝난 시점에서 Cars.isSize()
    - 메소드 명이 적절한가? isXXX()보다 matchXXX등이 맞을 수도 있다.
    - 이 메서드는 테스트에서만 사용되는데 이 메서드를 꼭 사용해서 테스트해야할까?
- Winners가 CarName을 가지는 것과 Car를 가지는 것 중 어느것이 더 좋을까?
- checkedException/uncheckedException의 차이에 대해 고민해보기
- move의 테스트를 generateRandomNumber() 때문에 하지 못하고 있는데 해결방법은 없는가?(전략패턴, 인터페이스로 수정)
    - 테스트에서만 사용되는 클래스는 test source set에 선언해주면 어떨까요?
- 테스트의 메소드 명을 모두 영어로 수정한다.
- 테스트에서 중복되는 것들을 Parameterized Test를 이용하여 리팩토링할 수 있지 않을까?
- 도메인에서 DTO가 맡은 역할이 무엇일까요?
    - DTO들이 필드를 하나씩만 가지고 있는데, 그래도 DTO를 쓸 필요가 있을까요?
- Util 클래스는 어떤 책임을 가져야 할까요?
- 어떤 예외를 던지는지 다시 확인해보기
    - 파라미터가 없는 메소드에서 `IllegalArgumentException`이 사용되는 것이 적절한가? 등등
- 변수, 클래스, 메서드 명을 전부 다 확인하여 의도에 맞게, 규칙에 맞게 지어졌는지 다시 확인해보기
- 오류 메시지를 가진 enum이 과연 필요한 것인가? 왜 오류 메시지를 사용하는 곳이 아니라 별도의 enum으로 분리했을까요?
- 도메인 객체가 모두 `vo` 패키지 내에 있는데 이것들이 정말 VO가 맞는지도 고민해보면 좋겠습니다.
- CarsTest에서 회차마다 입력한 자동차의 이름이 결과에 포함되어 있는지를 검증하지만 실제로 각 car들이 이동했는지를 검증하지 않습니다. 이 부분도 테스트하려면 어떻게 해야 할까요?

아래 README는 제 생각 적어둔거에요
https://github.com/OzRagwort/java-racingcar/blob/step2/docs/README.md

**Comment by seong-wooo** [comment]
저 하나 궁금한게 알린이 데이터 불변을 위해 dto를 사용한다고 하셨는데, 알린 코드중에 ResponseWinnersDto 클래스 getWinners 메서드를 써서 List<Car>를 전달받으면 컨트롤러에서도 전달받은 데이터를 수정할 수도 있는거 아니에요?

**Reply by OzRagwort** [reply]
제가 이해한 DTO을 이용해서 데이터의 불변을 유지한다는 의미는 DTO에서 get을 받아 사용하기 전까지로 이해했어요

그래서 만약 컨트롤러에서 get한 다음 데이터를 수정하고 다시 view로 보낸다면 DTO를 잘못사용하고 있는것같아요. 컨트롤러에서 수정을 해야한다면 꼭 컨트롤러에서 수정해야하는지, 도메인에서 그 로직을 할 수는 없는지 고민해봐야 할 것 같아요

다른 수정이 생길 수 있는 경우는 DTO에 참조형식으로 데이터를 넣고 DTO를 보내기 전, View에서 사용하기 전에 도메인에서 그 값을 수정하는 경우가 있을 수 있을 것 같아요. 

이런 사이드 이펙트를 방지하기위해 VO는 불변해야한다는 특징이 있더라구요. 일급 컬렉션이 조금 애매한데 저는 `Car`, `CarName`, `Position`, `Attempt`는 완전 불변하게 만들고 일급 컬렉션인 `Cars`는 final을 이용하여 재할당을 방지하고 getWinners등의 메서드에서는 `Collectors.unmodifiableList()`로 데이터가 수정되지 않도록 해봤어요.

`ResponseWinnersDto`의 `getWinners()` 메서드로 반환된 List의 형식은 `Cars`에서 `Collectors.unmodifiableList()`로 반환된 변경되지 않는 List입니다. 이 List는 `add()`, `remove()`, `set()` 등이 되지 않아서 수정할 수 없어요.그리고 각각의 Car는 불변 객체이기 때문에 수정할 수 없어요. 그렇기 때문에 사이드 이펙트는 발생하지 않을 것 같아요. 

저도 이부분에서 고민을 해본게 있는데 지금은 불변 객채와 수정 불가능 List를 이용했지만 결국 DTO가 도메인의 실제 데이터를 참조하고는 있는것 같아요. 그래서 DTO 생성자에서 `new ArrayList<>();`등으로 도메인과 참조를 끊어버리는 방법도 생각해봤었어요. 배카라는 어떻게 생각하시나요??

**Reply by seong-wooo** [reply]
`Collectors.unmodifiableList()`가 있는 지 몰랐네요. 

`new ArrayList<>()`를 사용한다는 것은 

```
  public List<RoundResult> getResults() {
        return new ArrayList<>(results);
    }
```

이런식으로 하신다는건가요 ?

제가 dto를 잘 모르긴하는데 괜찮은거 같으면서도 만약 result 가 linkedList 형태로 바뀐다면 수정하기 귀찮겠다라는 생각이 들어요

그리고 글 보면서 또 궁금해진게 Car를 vo 형식으로 구현한게 조금 어색한거같아요.

```
    public Car move(int number) {
        if (canMove(number)) {
            return new Car(name, position.increase());
        }
        return this;
    }
```
Car 객체가 새로운 Car 객체를 생성하는게 약간 어색해요. 
Car가 이동을 하는 메소드인데 이동할 때마다 새로운 Car를 반환하는 거니까 
제가 안 어색한데 어색하게 느끼는 걸 수도 있습니다ㅋㅋ 

날카로운 질문이었다면 따봉

**Reply by OzRagwort** [reply]
Car 객체가 매번 새로운 객체를 만드는건 저도 어색하다고 느껴요. 그래서 네오한테도 물어봤는데 상황에 따라 달라질것같다고 하더라구요.

상황에 따라 달라질거같다고 하는것은 그 객체의 상태가 얼마나 변화할지 이야기하는것같았어요. Position같이 자주 변화되는 10000번 이동하면 10000번 Position 객체가 생성되고 덕분에 Car도 10000번 생성이 되는데 이게 과연 효율적인게 맞느냐 하는 이야기니까요.

이 부분은 다같이 이야기 해봐도 좋을것같아요. 전 페어랑도 이 이야기를 해봤는데 결론이 아직 안났었어요. 

그런데 다른 분들의 피드백을 보면 어떤 방법이 있는것 같기도하고... ㅋㅋㅋㅋ

**Reply by seong-wooo** [reply]
제 생각은 position이랑 carName까지만 vo로 만드는게 좋을 것 같습니다.

**Reply by OzRagwort** [reply]
그 이유는 무엇인가요?

**Reply by seong-wooo** [reply]
car가 vo 형태라면 name과 position이 같은 car들은 전부 동일한 car를 의미해서?
이름과 포지션이 같지만 다른 car들도 존재할 수 있잖아여

**Reply by OzRagwort** [reply]
불변이라고 해서 동일하다는 의미는 아니지 않을까요? 

이름과 포지션이 같은 상태인 다른 객체가 있을 수 있을 것 같아요. equals는 true지만 ==는 false인 객체는 있을 수 있을 것 같아요.

`car가 vo 형태라면 name과 position이 같은 car들은 전부 동일한 car를 의미해서?`가 맞다고 하면, 만약 Car A와 Car B의 position이 둘다 0일때 같은 같은 객체를 참조하고 있는 건가요?

**Reply by seong-wooo** [reply]
저는 value object도 값이 같으면 동일해야한다고 생각해요.
예를 들어서 
```
String a1 = "a"
String a2 = "a"
```
이런 식으로 두 객체 모두 같은 값을 가리킬 때 a1 과 a2는 동일하다고 할 수 있잖아요
값 객체도 값의 성질을 가지고 있으니까 동일하게 적용할 수 있지 않을까요 ?
그리고 알린 코드 보면 hashCode() 재정의 하셨는데 그럼 == 또한 true로 나오지 않나여?

**Reply by OzRagwort** [reply]
String은 조금 특별한 케이스인것 같아요. 네오가 문자열은 비용이 비싸고 많이쓰여서 캐싱을 한다고 한것같은데 그래서 그런건지 String은 저렇게 해도 ==은 true가 나오더라구요. 

하지만 저희가 만들어서 사용하는 객체들은 String처럼 동작하지 않는 것 같아요. new로 새롭게 생성한다면 상태가 같아도 ==은 다른 것 같아요. 동등하지만 동일하지는 않다...? 저는 동등은 equals가 true인것 동일은 ==이 true인것으로 이해했어요

아래 코드로 예를 들면 더 쉬울것 같아요. 여기서 s1 == s2는 true일까요?
```
String s1 = new String("a");
String s2 = new String("a");
```
